lib/element.ts

```diff
@@ -0,0 +1,1055 @@
+   *
+   * The ElementArrayFinder is used to set up a chain of conditions that identify
+   * an array of elements. In particular, you can call all(locator) and
+   * filter(filterFn) to return a new ElementArrayFinder modified by the
+   * conditions, and you can call get(index) to return a single ElementFinder at
+   * position 'index'.
+   *
+   * Similar to jquery, ElementArrayFinder will search all branches of the DOM
+   * to find the elements that satisfy the conditions (i.e. all, filter, get).
+   * However, an ElementArrayFinder will not actually retrieve the elements until
+   * an action is called, which means it can be set up in helper files (i.e.
+   * page objects) before the page is available, and reused as the page changes.
+   *
+   * You can treat an ElementArrayFinder as an array of WebElements for most
+   * purposes, in particular, you may perform actions (i.e. click, getText) on
+   * them as you would an array of WebElements. The action will apply to
+   * every element identified by the ElementArrayFinder. ElementArrayFinder
+   * extends Promise, and once an action is performed on an ElementArrayFinder,
+   * the latest result can be accessed using then, and will be returned as an
+   * array of the results; the array has length equal to the length of the
+   * elements found by the ElementArrayFinder and each result represents the
+   * result of performing the action on the element. Unlike a WebElement, an
+   * ElementArrayFinder will wait for the angular app to settle before
+   * performing finds or actions.
+   *
+   * @alias element.all(locator)
+   * @view
+   * <ul class="items">
+   *   <li>First</li>
+   *   <li>Second</li>
+   *   <li>Third</li>
+   * </ul>
+   *
+   * @example
+   * element.all(by.css('.items li')).then(function(items) {
+   *   expect(items.length).toBe(3);
+   *   expect(items[0].getText()).toBe('First');
+   * });
+   *
+   * @constructor
+   * @param {Protractor} ptor A protractor instance.
+   * @param {function(): Array.<webdriver.WebElement>} getWebElements A function
+   *    that returns a list of the underlying Web Elements.
+   * @param {webdriver.Locator} locator The most relevant locator. It is only
+   *    used for error reporting and ElementArrayFinder.locator.
+   * @param {Array.<webdriver.promise.Promise>} opt_actionResults An array
+   *    of promises which will be retrieved with then. Resolves to the latest
+   *    action result, or null if no action has been called.
+   * @return {ElementArrayFinder}
+   */
+  constructor(
+      private ptor_: any, getWebElements: Function, private locator_: any,
+      public actionResults_: webdriver.Promise = null) {
+    this.getWebElements = getWebElements || null;
+
+    // TODO(juliemr): might it be easier to combine this with our docs and just
+    // wrap each
+    // one explicity with its own documentation?
+    WEB_ELEMENT_FUNCTIONS.forEach((fnName: string) => {
+      (<any>this)[fnName] = function() {
```

Comment:
When I tried to do that here, I got errors below because you can't use `arguments` in a fat-arrow function.